### PR TITLE
tests: drivers: counter: Fix test for nrf51

### LINKED ISSUE #24369: tests/drivers/counter/counter_basic_api failure on nRF51-DK
**Describe the bug**

While getting baseline test results for nRF counter APIs in order to test changes related to the new devicetree.h, I noticed counter_basic_api is failing for nRF51-DK.

**To Reproduce**

I'm using sanitycheck's [device testing](https://docs.zephyrproject.org/latest/guides/test/sanitycheck.html#running-tests-on-hardware) feature to test nRF51-DK, nRF52840-DK, nRF52-DK like so:

```
./scripts/sanitycheck --device-testing \
		      --hardware-map ~/zp/hw-map.yaml \
		      -T tests/drivers/counter \
		      --ninja
```

My [hw-map.yaml file](https://gist.github.com/mbolivar-nordic/329018e5be3778cba00f7ec642b6e2e2) associates devices with serial ports and nrfjprog IDs. 

Test results (with #24350 applied to satisfy the platform whitelist for nRF52840-DK) show the above mentioned failure. It is repeatable. Other tests always seem to pass.

**Expected behavior**

All tests should pass.

**Impact**

Unclear. Do we need this test to pass on this board?

**Screenshots or console output**

```
$ ./scripts/sanitycheck --device-testing  --hardware-map ~/zp/hw-map.yaml  -T tests/drivers/counter  --ninja
Renaming output directory to /home/mbolivar/zp/zephyr/sanity-out.4
INFO    - JOBS: 8
INFO    - Building initial testcase list...
INFO    - 4 test configurations selected, 5 configurations discarded due to filters.

Device testing on:

| Platform            |           ID | Serial device                                         |
|---------------------|--------------|-------------------------------------------------------|
| nrf51dk_nrf51422    | 000681986260 | /dev/serial/by-id/usb-SEGGER_J-Link_000681986260-if00 |
| nrf52840dk_nrf52840 | 000683441799 | /dev/serial/by-id/usb-SEGGER_J-Link_000683441799-if00 |
| nrf52dk_nrf52832    | 000682296740 | /dev/serial/by-id/usb-SEGGER_J-Link_000682296740-if00 |

INFO    - Adding tasks to the queue...
INFO    - Total complete:    3/   4  75%  skipped:    0, failed:    0
ERROR   - nrf51dk_nrf51422          tests/drivers/counter/counter_basic_api/drivers.counter FAILED: Failed
ERROR   - see: /home/mbolivar/zp/zephyr/sanity-out/nrf51dk_nrf51422/tests/drivers/counter/counter_basic_api/drivers.counter/handler.log
INFO    - Total complete:    4/   4  100%  skipped:    0, failed:    1
INFO    - 3 of 4 tests passed (75.00%), 1 failed, 0 skipped with 0 warnings in 30.41 seconds
INFO    - In total 3 test cases were executed on 3 out of total 241 platforms (1.24%)

Hardware distribution summary:

| Board               |           ID |   Counter |
|---------------------|--------------|-----------|
| nrf52dk_nrf52832    | 000682296740 |         1 |
| nrf51dk_nrf51422    | 000681986260 |         1 |
| nrf52840dk_nrf52840 | 000683441799 |         2 |
```

Failing log:

```
$ cat /home/mbolivar/zp/zephyr/sanity-out/nrf51dk_nrf51422/tests/drivers/counter/counter_basic_api/drivers.counter/handler.log
*** Booting Zephyr OS build zephyr-v2.2.0-1473-gd87c4556d878  ***
Running test suite test_counter
===================================================================
starting test - test_set_top_value_with_alarm
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_set_top_value_with_alarm
===================================================================
starting test - test_single_shot_alarm_notop
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_single_shot_alarm_notop
===================================================================
starting test - test_single_shot_alarm_top
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:265: alarm_handler: (now >= counter is false)
RTC_0: Alarm (655) too early now: 0 (counting up).
FAIL - test_single_shot_alarm_top
===================================================================
starting test - test_multiple_alarms
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_multiple_alarms
===================================================================
starting test - test_all_channels
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_all_channels
===================================================================
starting test - test_late_alarm
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_late_alarm
===================================================================
starting test - test_late_alarm_error
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_late_alarm_error
===================================================================
starting test - test_short_relative_alarm
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_short_relative_alarm
===================================================================
starting test - test_cancelled_alarm_does_not_expire
Testing TIMER_0
Testing TIMER_1
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:802: test_cancelled_alarm_does_not_expire_instance: (0 not equal to err)
TIMER_1: Failed to set an alarm (err: -22)
FAIL - test_cancelled_alarm_does_not_expire
===================================================================
starting test - test_set_top_value_without_alarm
Testing TIMER_0
Testing TIMER_1
Testing TIMER_2
Testing RTC_0
Testing RTC_0
PASS - test_set_top_value_without_alarm
===================================================================
Test suite test_counter failed.
===================================================================
PROJECT EXECUTION FAILED
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA: 5a149dfca6 plus #24350 applied to satisfy the platform whitelist for nRF52840-DK

### PR DESCRIPTION
**Pull request by nordic-krch**
Fixes test which did not take into account that counter may
wrap (nrf51 has 16bit timers) and was setting alarm to wrong
value. Error was not seen on platforms with higher top value.

Fixes #24369.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 1]
```diff
@@ -797,7 +799,8 @@ static void test_cancelled_alarm_does_not_expire_instance(const char *dev_name)
 
 		k_busy_wait(us/2 + i);
 
-		alarm_cfg.ticks = alarm_cfg.ticks + 2*alarm_cfg.ticks;
```
Comment (nordic-krch) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:800]: test was suppose to set some alarm in the future (it is cancelled anyway) and for non wrapping (within test) counters this formula was ok but it was actually wrong and initial intention was that it is `2*ticks` (2 ms from now).

Comment (anangl) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:800]: > and initial intention was that it is `2*ticks` (2 ms from now).

The commit message would be a better place for such additional explanation.

**Review by anangl — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
The test is still failing for me. https://gist.github.com/mbolivar-nordic/8410c20dda61345b67e806d340aa6c0c

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
-1 since this is not a fix in my testing

**Comment by nordic-krch** [comment]
>-1 since this is not a fix in my testing

That test was spurious as it was failing only when alarm expired when counter wrapped. That was revealing a bug in the test - validation that alarm came not earlier than requested. Fixed.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 2]
```diff
@@ -254,23 +254,33 @@ void test_set_top_value_without_alarm(void)
 static void alarm_handler(struct device *dev, u8_t chan_id, u32_t counter,
 			  void *user_data)
 {
+	/* Arbitrary limit for alarm processing - time between hw expiration
+	 * and read-out from counter in the handler.
+	 */
+	static const u64_t processing_limit_us = 1000;
```
Comment (mbolivar-nordic) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:260]: Why static? Just curious.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 3]
```diff
@@ -254,23 +254,33 @@ void test_set_top_value_without_alarm(void)
...
-		zassert_true(now <= counter,
-			"%s: Alarm (%d) too early now: %d (counting down).",
-			dev->config->name, counter, now);
+		diff = (now > counter) ?
+			(counter + top - now) : (counter - now);
 	}
 
+	zassert_true(diff < counter_us_to_ticks(dev, processing_limit_us),
+			"Unexpected distance between reported alarm value(%u) "
+			"and actual counter value (%u), top:%d",
```
Comment (mbolivar-nordic) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:281]: Would be good to include processing_limit_us in the output here too, I think.

**Review by mbolivar-nordic — APPROVED** [review]
Test is passing now so I'm happy, thanks! A couple of non-blocking style things.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c (hunk 2 above)
Comment (nordic-krch) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:260]: puts variable in flash and not on stack.

**Comment by nordic-krch** [comment]
i'm getting counter test failures on `mps2_an385`. However, locally those tests are passing. @nashif do you know who maintains counter for `mps_an385`?

**Comment by carlescufi** [comment]
@nordic-krch there seems to be a regression in this PR: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/69627/3/console

I have re-run CI but the failure re-appeared so this is not a false positive.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c (hunk 2 above)
Comment (mbolivar-nordic) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:260]: Are we really so close to blowing the stack on this simple test case? I'm not objecting to it, it just looked weird.

**Comment by nordic-krch** [comment]
@carlescufi yes, on CI is it consistent but on my setup it is not. This test is passing on my machine and changes do not touch failing test so i think it is also failing on master.

**Comment by nordic-krch** [comment]
created bug report for failing test: #24635

**Comment by nordic-krch** [comment]
@carlescufi can you merge? We accidentally got green build. Root cause is known. #22904 should eventually fix it.

**Pull request merged**